### PR TITLE
Java: Add binding between annotation and sink-param in MyBatis SQL Injection query

### PR DESCRIPTION
**Pull request by ka1n4t**
Add binding between annotation and sink-param.

In below case, the rule `MyBatisAnnotationSqlInjection.ql` will find a vulnerability at `parameter name` in function `testParamAnno1()`. That is false positive.

```java
@Select("select * from user_info where name = #{name} and age = ${age}")
String testParamAnno1(@Param("name") String name, Integer age);

@Select("select * from user_info where age = #{age}")
String testParamAnno2(@Param("age") String age);
```

The root cause is, `isMybatisXmlOrAnnotationSqlInjection()` find `annotation @Param` in `testParamAnno2()` and bind it to `${age}` in `@Select("select * from user_info where name = #{name} and age = ${age}")` when processing `parameter name in testParamAnno1()`. So it raise a vulnerability to `name` when it saw `${age}` binding with `@Param("age")`.

### CODE REVIEW

**Comment by atorralba** [comment]
This looks like an interesting addition, thanks for the contribution! Would you mind adding a test case highlighting the issue you're fixing?

**Comment by ka1n4t** [comment]
> This looks like an interesting addition, thanks for the contribution! Would you mind adding a test case highlighting the issue you're fixing?

Of course! I have committed the test case

File: java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll [hunk 1]
```diff
@@ -134,7 +134,8 @@ predicate isMybatisXmlOrAnnotationSqlInjection(
           .matches("${" + annotation.getValue("value").(CompileTimeConstantExpr).getStringValue() +
               "%}") and
       annotation.getType() instanceof TypeParam and
-      ma.getAnArgument() = node.asExpr()
+      ma.getAnArgument() = node.asExpr() and
+      annotation.getTarget() = ma.getMethod().getParameter(node.asExpr().getIndex())
```
Comment (atorralba) [inline java/ql/src/experimental/Security/CWE/CWE-089/MyBatisCommonLib.qll:138]: Just a small change to clarify the intention here:

```suggestion
      annotation.getTarget() = ma.getMethod().getParameter(node.asExpr().(Argument).getParameterPos())
```

File: java/ql/test/experimental/query-tests/security/CWE-089/src/main/MybatisSqlInjection.java [hunk 2]
```diff
@@ -89,4 +89,14 @@ public void badUpdate(@RequestParam String name) {
 	public void badInsert(@RequestParam String name) {
 		mybatisSqlInjectionService.badInsert(name);
 	}
+
+	@GetMapping(value = "kkbad1")
+	public void kkbad1(@RequestParam String name, @RequestParam Integer age) {
+		mybatisSqlInjectionService.kkbad1(name, age);
```
Comment (atorralba) [inline java/ql/test/experimental/query-tests/security/CWE-089/src/main/MybatisSqlInjection.java:95]:
```suggestion
	@GetMapping(value = "good2")
	public void good2(@RequestParam String name, @RequestParam Integer age) {
		mybatisSqlInjectionService.good2(name, age);
```

File: java/ql/test/experimental/query-tests/security/CWE-089/src/main/MybatisSqlInjection.java [hunk 3]
```diff
@@ -89,4 +89,14 @@ public void badUpdate(@RequestParam String name) {
...
+	}
+
+	@GetMapping(value = "kkbad2")
+	public void kkbad2(@RequestParam String age) {
+		mybatisSqlInjectionService.kkbad2(age);
```
Comment (atorralba) [inline java/ql/test/experimental/query-tests/security/CWE-089/src/main/MybatisSqlInjection.java:100]:
```suggestion
	@GetMapping(value = "good3")
	public void good3(@RequestParam String age) {
		mybatisSqlInjectionService.good3(age);
```

File: java/ql/test/experimental/query-tests/security/CWE-089/src/main/MybatisSqlInjectionService.java [hunk 4]
```diff
@@ -72,4 +72,12 @@ public void badUpdate(String input) {
 	public void badInsert(String input) {
 		sqlInjectionMapper.badInsert(input);
 	}
+
+	public void kkbad1(String name, Integer age){
+		sqlInjectionMapper.kkbad1(name, age);
```
Comment (atorralba) [inline java/ql/test/experimental/query-tests/security/CWE-089/src/main/MybatisSqlInjectionService.java:77]:
```suggestion
	public void good2(String name, Integer age){
		sqlInjectionMapper.good2(name, age);
```

File: java/ql/test/experimental/query-tests/security/CWE-089/src/main/MybatisSqlInjectionService.java [hunk 5]
```diff
@@ -72,4 +72,12 @@ public void badUpdate(String input) {
...
+	}
+
+	public void kkbad2(String age){
+		sqlInjectionMapper.kkbad2(age);
```
Comment (atorralba) [inline java/ql/test/experimental/query-tests/security/CWE-089/src/main/MybatisSqlInjectionService.java:81]:
```suggestion
	public void good3(String age){
		sqlInjectionMapper.good3(age);
```

File: java/ql/test/experimental/query-tests/security/CWE-089/src/main/SqlInjectionMapper.java [hunk 6]
```diff
@@ -37,26 +37,33 @@ public interface SqlInjectionMapper {
...
 	@InsertProvider(
-		type = MyBatisProvider.class,
-		method = "badInsert"
+			type = MyBatisProvider.class,
+			method = "badInsert"
 	)
 	void badInsert(String input);
+
+	@Select("select * from user_info where name = #{name} and age = ${age}")
+	String kkbad1(@Param("name") String name, Integer age);
```
Comment (atorralba) [inline java/ql/test/experimental/query-tests/security/CWE-089/src/main/SqlInjectionMapper.java:64]:
```suggestion
	String good2(@Param("name") String name, Integer age);
```

File: java/ql/test/experimental/query-tests/security/CWE-089/src/main/SqlInjectionMapper.java [hunk 7]
```diff
@@ -37,26 +37,33 @@ public interface SqlInjectionMapper {
...
+
+	@Select("select * from user_info where age = #{age}")
+	String kkbad2(@Param("age") String age);
```
Comment (atorralba) [inline java/ql/test/experimental/query-tests/security/CWE-089/src/main/SqlInjectionMapper.java:67]:
```suggestion
	String good3(@Param("age") String age);
```

**Review by atorralba — COMMENTED** [review]
Thanks for adding the test! I added some small suggestions to improve readability, but otherwise this LGTM.

**Review by atorralba — APPROVED** [review]

**Review by atorralba — APPROVED** [review]

**Pull request merged**